test/meta.cpp

```diff
@@ -851,7 +955,7 @@ TEST_F(Meta, MetaDataConstStatic) {
 
     ASSERT_TRUE(data);
     ASSERT_EQ(data.parent(), meta::resolve("data"));
-    ASSERT_EQ(data.type(), meta::resolve<int>());
+    ASSERT_EQ(data.type(), meta::resolve<const int>());
```

Comment:
Finally, I got a pause from other projects and I **want** to dedicate some spare time to `meta` so as to add full support for references. I find it useful to move back and forth big types actually, so I think it's worth it.  
Unfortunately, I tried this change on a software of mine and found that it added as much as it removed. In particular, the fact that a `const int` data member is no longer identified as `int` from the meta system breaks some code my side. Similarly, the same applies to other types and different qualifiers.  

From my point of view, the _opacity_ of the meta system with regards to types was a value, not something to be defeated.  
Imagine an editor that works with meta types. Where I had to deal only with `int`, `double` and so on, now I've to deal with `int`, `const int`, `double`, `const double`, and so on. Not really an improvement.

Isn't there a way to do that à la rttr, using `std::reference_wrapper`s where explicitly required?  
During the last days I thought also to something more _C++17-ish_, where I can specify a function type or a data type aside the actual member. As an example:

```
meta::reflect<my_type>("T").data<&my_type::i, std::reference_wrapper>("i");
// as well as
meta::reflect<my_type>("T").data<&my_type::i, double>("i");
```
In other terms, a way to tell to the system - _ok, this is the data member, no matter what its type is, return it to me using the given type_.
This could work for data members as well as functions, constructors and so on, other than for pointers-to-references conversion under the hood when you say - _here is a pointer, return it as a reference, please_.

What do you think about?

Reply:
Yeah, the problem with this PR is exactly that it is not quite backwards compatible as it was. But after #4 it should be able to work in most cases as-is, since we can make `T&` funcs iterable too:

```
// syntax might not be 100% accurate since i did this from the top of my head
auto r = func.invoke(...); // returns a T&
r.type().data("value").value(r); // should work with T, T& or T* (#4)
```

I think it is important to know the const-ness information, since the behaviour of a const object is a bit different than a non-const in some situations. Otherwise we would have to always perform a `const_cast` (which is undefined behaviour in some instances, see below). We would also say goodbye to any opportunity of overload resolution.

```
static const int n = 0;

struct C {
  const int& n() { return n; }
  void test(int& n) { n++ };
}

meta::func func_n = ...;
meta::func func_test = ...;

meta::any c = c; // is of type C
meta::any r = func_n.invoke(c);
func_test.invoke(c, r);
```

This (the last `invoke`) is UB, since a const_cast will need to be performed implicitly and the standard allows the compiler to allocate the global `n` in a read-only section in the executable. In the optimal case things crash. We also cannot do any implicit copy because we don't know if it is const or not.

I dislike the `rttr` approach with regards to references. Simply because we can (and IMO should) use the language itself to to deduce the return type. If the function returns a reference, I would like a reference to the returned object. In the (unlikely) case that I want a value, only then I should request it explicitly. Even better would be something like:

```meta::func f; 
auto ret = f.invoke(...); // returns a &
ret.to_value(); // explicitly copies the object -- could be simply a call to ret.conv<T>() since (const) T& is convertible to a value
```

This approach would also make more difficult to use Clang-based tools on the source code and generate bindings automatically, because then we need to provide extra context to the parser.

Also, I think part of what you are proposing at the end is that #4 does, but with pointer types instead. I think we should start there first, since that change is fully backwards compatible and will significantly improve this PR too.